### PR TITLE
ADEN-9078 InstantConfig interpreter

### PR DESCRIPTION
**Pull request by Bielik20**

### CODE REVIEW

File: spec/ad-services/instant-config/instant-config.interpreter.spec.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		};
+
+		mockResponses([true, true, true, true]);
+		interpreter.getValues(instantConfig);
+
+		expect(browserIsValidStub.firstCall.args[0]).to.equal(instantConfig.babDetection[0].browsers);
+		expect(deviceIsValidStub.firstCall.args[0]).to.equal(instantConfig.babDetection[0].devices);
+		expect(domainIsValidStub.firstCall.args[0]).to.equal(instantConfig.babDetection[0].domains);
+		expect(regionIsValidStub.firstCall.args[0]).to.equal(instantConfig.babDetection[0]);
+		expect(regionIsValidStub.firstCall.args[1]).to.equal('babDetection-0');
```
Comment (fraszczakszymon) [inline spec/ad-services/instant-config/instant-config.interpreter.spec.ts]: We need to check tracking and confirm with Martyna/Dorota whether they are fine with new labels (with additional group id)

File: spec/ad-services/instant-config/instant-config.loader.spec.ts [hunk 2]
```diff
@@ -1,9 +1,11 @@
+import { instantConfigLoader } from '@wikia/ad-services/instant-config/instant-config.loader';
```
Comment (fraszczakszymon) [inline spec/ad-services/instant-config/instant-config.loader.spec.ts]: When did we started this kind of pattern? Why not `instant-config-loader`?

File: spec/ad-services/instant-config/matchers/device-matcher.spec.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import { utils } from '@wikia/ad-engine';
+import { DeviceMatcher } from '@wikia/ad-services/instant-config/matchers/device-matcher';
+import { expect } from 'chai';
+import * as sinon from 'sinon';
+
+describe('Device Matcher', () => {
```
Comment (fraszczakszymon) [inline spec/ad-services/instant-config/matchers/device-matcher.spec.ts]: :gem: Good scenarios

File: spec/ad-services/instant-config/matchers/browser-matcher.spec.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import { utils } from '@wikia/ad-engine';
+import { BrowserMatcher } from '@wikia/ad-services/instant-config/matchers/browser-matcher';
+import { expect } from 'chai';
+import * as sinon from 'sinon';
+
+describe('Browser Matcher', () => {
```
Comment (fraszczakszymon) [inline spec/ad-services/instant-config/matchers/browser-matcher.spec.ts]: :gem: Good scenarios

File: spec/ad-services/instant-config/matchers/domain-matcher.spec.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import { DomainMatcher } from '@wikia/ad-services/instant-config/matchers/domain-matcher';
+import { expect } from 'chai';
+
+describe('Domain Matcher', () => {
```
Comment (fraszczakszymon) [inline spec/ad-services/instant-config/matchers/domain-matcher.spec.ts]: :gem: Good scenarios

File: spec/ad-services/instant-config/matchers/region-matcher.spec.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
+import { utils } from '@wikia/ad-engine';
+import { RegionMatcher } from '@wikia/ad-services/instant-config/matchers/region-matcher';
+import { expect } from 'chai';
+import * as sinon from 'sinon';
+
+describe('Region Matcher', () => {
```
Comment (fraszczakszymon) [inline spec/ad-services/instant-config/matchers/region-matcher.spec.ts]: :gem: Good scenarios

File: spec/ad-services/instant-config/matchers/region-matcher.spec.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+	it('should filter out invalid', () => {
+		regionMatcher.isValid({
+			regions: ['XX/25-cached', 'US-cached', 'PL/25', 'non-PL', 'UK'],
+		});
+
+		expect(isProperGeoStub.getCall(0).args[0]).to.deep.equal(['non-PL', 'UK']);
+	});
+
+	it('should apply sampling', () => {
```
Comment (fraszczakszymon) [inline spec/ad-services/instant-config/matchers/region-matcher.spec.ts]: Unfortunately, I believe we have to move sampling from geo module and apply it for whole feature flag groups.

File: src/ad-services/instant-config/instant-config.service.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+		return defaultValue;
+	}
+
+	/**
+	 * Use only for legacy wgAdDriver keys
+	 * @deprecated
+	 */
+	isGeoEnabled(key: string): boolean {
+		if (!key.startsWith('wgAdDriver')) {
+			throw new Error('This method should be only used for legacy wgAdDriver keys');
```
Comment (fraszczakszymon) [inline src/ad-services/instant-config/instant-config.service.ts]: 💎

File: src/ad-services/instant-config/matchers/region-matcher.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
+import { utils } from '@ad-engine/core';
+import { cacheSuffix, InstantConfigGroup, samplingSeparator } from '../instant-config.models';
+
+// TODO: Move sampling and cache on top of the group. Make so that cache works per group.
```
Comment (fraszczakszymon) [inline src/ad-services/instant-config/matchers/region-matcher.ts]: :+1:

File: src/ad-engine/services/local-cache.ts [hunk 10]
```diff
@@ -1,13 +1,13 @@
...
+import { UniversalStorage } from './universal-storage';
 
 interface CacheItem<T = any> {
 	expires?: number;
 	data: T;
 }
 
 class LocalCache {
-	private storage = new LocalStorage();
+	private storage = new UniversalStorage();
```
Comment (fraszczakszymon) [inline src/ad-engine/services/local-cache.ts]: What does `UniversalStorage` mean? Why do we change it while leaving `LocalCache` as "local"?

Comment (Bielik20) [reply src/ad-engine/services/local-cache.ts]: Damian started using it for session cookie by adding argument to constructor. Thus it stopped being LocalStorage per se. It became something more generic/universal - hence the new name.

Although I think it should have reversed control. Meaning Session storage should use universal one and expose methods through its API, not the other way around. But I am afraid it is beyond scope of this ticket.

File: spec/ad-services/instant-config/instant-config.loader.spec.ts (hunk 2 above)
Comment (Bielik20) [reply spec/ad-services/instant-config/instant-config.loader.spec.ts]: I am not very attached to this. I just wanted something to distinguish certain parts of instant-config. That's why it is always `instant-config` followed by `.` and specific responsibility. But I can change it if you feel like it's better to stick to `-` everywhere.

File: src/ad-engine/services/local-cache.ts (hunk 10 above)
Comment (fraszczakszymon) [reply src/ad-engine/services/local-cache.ts]: God bless you (for the last sentence) 😂

**Pull request merged**